components/lib/state/src/main/java/mozilla/components/lib/state/Store.kt

```diff
@@ -0,0 +1,108 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.lib.state
+
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.asCoroutineDispatcher
+import kotlinx.coroutines.launch
+import java.util.concurrent.Executors
+
+typealias Reducer<S, A> = (S, A) -> S
+typealias Observer<S> = (S) -> Unit
+
+/**
+ * A generic store holding an immutable [State].
+ *
+ * The [State] can only be modified by dispatching [Action]s which will create a new state and notify all registered
+ * [Observer]s.
+ *
+ * @param initialState The initial state until a dispatched [Action] creates a new state.
+ * @param reducer A function that gets the current [State] and [Action] passed in and will return a new [State].
+ */
+open class Store<S : State, A : Action>(
+    initialState: S,
+    private val reducer: Reducer<S, A>
+) {
+    private val storeContext = Executors.newSingleThreadExecutor().asCoroutineDispatcher()
+    private val storeScope = CoroutineScope(storeContext)
+    private val subscriptions: MutableList<Subscription<S, A>> = mutableListOf()
+    private var currentState = initialState
+
+    /**
+     * The current [State].
+     */
+    val state: S
+        get() = currentState
+
+    /**
+     * Registers an [Observer] function that will be invoked whenever the [State] changes.
+     *
+     * @param receiveInitialState If true the observer function will be invoked immediately with the current state.
+     * @return A [Subscription] object that can be used to unsubscribe from further state changes.
+     */
+    fun observe(receiveInitialState: Boolean = true, observer: Observer<S>): Subscription<S, A> {
+        val subscription = Subscription(observer, store = this)
+
+        synchronized(subscriptions) {
+            subscriptions.add(subscription)
+        }
+
+        if (receiveInitialState) {
+            observer.invoke(currentState)
+        }
+
+        return subscription
```

Comment:
It's possible for the `subscription` to outlive the `store`, right? Lifetimes of these two aren't currently coordinated at all, once you return the `subscription`. And subscription has a reference to the `store`, which is opaque to users of the subscription. I don't think this sort of a class would actually compile in Rust, where this stuff is enforced.

Reply:
This is similar to how redux handles subscriptions, but using a `WeakRef` for the store might be better.